site/js/modules/webui/input.js

```diff
@@ -0,0 +1,187 @@
+import * as Autograder from '../autograder/base.js';
+
+// A general representation of a user input field.
+class Field {
+    constructor(
+            name, displayName,
+            {
+                type = 'text', underlyingType = 'string', required = false, placeholder = '',
+                inputClass = '', attributes = '', labelBefore = true, extractInputFunc = undefined
+            } = {}) {
+        // The name of the field which will be used for targeting the field.
+        this.name = name;
+
+        // The display name that will be shown to the user.
+        this.displayName = displayName;
+
+        // The HTML input type.
+        // Supports checkbox, email, number, password, and text types.
+        this.type = type;
+
+        // The underlying type of the field.
+        // Non-standard types are parsed to JSON.
+        // If a non-standard type cannot be parsed to JSON,
+        // a validation error is raised.
+        this.underlyingType = underlyingType;
```

Comment:
This seems strange to me.
I feel like these should probably be merged.

In what cases would we have an underlying type that does not implicitly choose the type of input element?
Seems like it would be more likely for a user to just set the wrong combo and get an error.

A user choosing the input element seems strange when they are also passing in an underlying type.

I think you should just know about a subset of types, handle those specially, and fallback the rest to JSON.